src/analysis_and_optimization/Optimize.ml

```diff
@@ -1017,62 +1017,181 @@ let block_fixing mir =
+ by some circumstance, and then updating the MIR with that set.
+
+    gen_variables: the variables that must be added to the set at the given statement
+    update_expr: update an MIR expression given the variable set
+    update_decl: update an MIR decl given the variable set
+    extra_variables: the set of variables that are implied to be in the set by a given variable in the set (usually empty, sometimes unrepresented variables like _in__ variables)
+    initial_variables: the initial known members of the set of variables
+    stmt: the MIR statement to optimize.
+*)
+let optimize_minimal_variables
+    ~(gen_variables :
+          (int, Stmt.Located.Non_recursive.t) Map.Poly.t
+       -> int
+       -> string Set.Poly.t
+       -> string Set.Poly.t)
+    ~(update_expr : string Set.Poly.t -> Expr.Typed.t -> Expr.Typed.t)
+    ~(update_decl :
+          bool
+       -> UnsizedType.autodifftype * 'a Type.t
+       -> UnsizedType.autodifftype * 'a Type.t)
+    ~(extra_variables : string -> string Set.Poly.t)
+    ~(initial_variables : string Set.Poly.t) (stmt : Stmt.Located.t) =
+  let rev_flowgraph, flowgraph_to_mir =
+    Monotone_framework.inverse_flowgraph_of_stmt stmt
+  in
+  let fwd_flowgraph = Monotone_framework.reverse rev_flowgraph in
+  let (module Rev_Flowgraph) = rev_flowgraph in
+  let (module Fwd_Flowgraph) = fwd_flowgraph in
+  let ad_levels =
+    Monotone_framework.minimal_variables_mfp
+      (module Fwd_Flowgraph)
+      (module Rev_Flowgraph)
+      flowgraph_to_mir initial_variables gen_variables
+  in
+  let optimize_min_vars_stmt_base i stmt =
+    let variable_set =
+      let exits = (Map.find_exn ad_levels i).exit in
+      Set.Poly.union exits (union_map exits ~f:extra_variables)
+    in
+    match
+      Stmt.Fixed.Pattern.map (update_expr variable_set) (fun x -> x) stmt
+    with
+    | Decl {decl_id; decl_type; decl_adtype} ->
+        let ad', type' =
+          update_decl (Set.mem variable_set decl_id) (decl_adtype, decl_type)
+        in
+        Stmt.Fixed.Pattern.Decl {decl_adtype= ad'; decl_id; decl_type= type'}
+    | s -> s
+  in
+  map_rec_stmt_loc_num flowgraph_to_mir optimize_min_vars_stmt_base
+    (Map.find_exn flowgraph_to_mir 1)
+
+(** A generic optimization pass for finding a minimal set of variables that are generated
+ *by some circumstance, and then updating the MIR with that set.
+ *
+ * gen_variables: the variables that must be added to the set at the given statement
+ * update_expr: update an MIR expression given the variable set
+ * update_stmt: update an MIR statement given the variable set
+ * extra_variables: the set of variables that are implied to be in the set by a given variable in the set (usually empty, sometimes unrepresented variables like _in__ variables)
+ * initial_variables: the initial known members of the set of variables
+ * stmt: the MIR statement to optimize.
+ *)
+let optimize_mem_pattern
```

Comment:
This is doing the same thing as `optimize_minimal_variables`, right? Can you use that?

Reply:
Oh I see that it's doing update_stmt instead of update_decl. I didn't think you'd need that, do you need to modify anything except for declarations and expressions?

Reply:
The only other place is in `Assignments` with the right hand side being `FnReadParam`.

```ocaml
  | Assignment
      ( ((name : string), (ut : UnsizedType.t), lhs)
      , { pattern=
            FunApp (CompilerInternal (FnReadParam (constrain_op, SoA)), args)
        ; meta= emeta } ) ->
      if Set.Poly.exists ~f:(fun x -> x = name) var_name then
        let query_index = mod_index mod_expr in
        Assignment
          ( (name, ut, List.map ~f:query_index lhs)
          , { pattern=
                FunApp
                  ( CompilerInternal (FnReadParam (constrain_op, AoS))
                  , List.map ~f:mod_expr args )
            ; meta= emeta } )
      else
        Assignment
          ( (name, ut, lhs)
          , { pattern=
                FunApp
                  (CompilerInternal (FnReadParam (constrain_op, SoA)), args)
            ; meta= emeta } )
```

I think the reason we need that is because the right hand side expression does not know the name of the object we are assigning to. We need to check the name and modify the deserialization read on the right hand side from to read out an array of struct instead of a struct of arrays ala

````cpp
deserializer.read<var_value<Eigen::Matrix<double>>>(...)
``` 
to 
```
deserialier.read<Eigen::Matrix<var>>(...)
```

when the name of the assignee is the same as one of the failed objects

The only other modified statement is for `NRFunApp`. Is the NR for `Non Returning`?

I'm finishing up the chapters from principles of programming analysis rn, am I reading right that I can delete everything else here that just recursively changes inner expressions? Can I also remove the paths that modify inner lists of statements for things like Block and For?

Reply:
Ah it does make sense that you'd need to modify assignments. But in that case we should generalize `optimize_minimal_variables` rather than repeating most of it

Yep, the NR is for non-returning